client/network/src/persist_peers.rs

```diff
@@ -0,0 +1,308 @@
+use std::{
+	collections::{HashMap, HashSet},
+	fmt,
+	future::Future,
+	io,
+	path::{Path, PathBuf},
+	pin::Pin,
+	sync::Arc,
+	task::{Context, Poll},
+	time::{Duration, Instant},
+};
+
+use futures::FutureExt;
+use lru::LruCache;
+
+use sc_peerset::PeersetHandle;
+
+use crate::{Multiaddr, PeerId};
+
+type BoxedFuture<T> = Pin<Box<dyn Future<Output = T> + Send + 'static>>;
+type Never = std::convert::Infallible;
+type ProtocolType = String;
+
+const FLUSH_INTERVAL: Duration = Duration::from_secs(5);
+const PEER_ADDRS_CACHE_SIZE: usize = 100;
+
+pub struct PersistPeerAddrs {
+	paths: Arc<Paths>,
+	flushed_at: Instant,
+	protocols: HashMap<ProtocolType, LruCache<PeerId, HashSet<Multiaddr>>>,
+	busy: Option<BoxedFuture<Result<(), io::Error>>>,
+}
+
+impl PersistPeerAddrs {
+	pub fn load(dir: impl AsRef<Path>) -> Self {
+		let paths = Paths::new(dir, "peer-addrs");
+
+		let protocols = match persist_peer_addrs::load(&paths.path) {
+			Ok(restored) => restored,
+			Err(reason) => {
+				log::warn!("Failed to load peer addresses: {:?}", reason);
+				Default::default()
+			},
+		};
+
+		let protocols = protocols
+			.into_iter()
+			.map(|(protocol, entries)| {
+				let cache = entries.into_iter().rev().fold(
+					LruCache::new(PEER_ADDRS_CACHE_SIZE),
+					|mut acc, persist_peer_addrs::PeerEntry { peer_id, addrs }| {
+						if let Ok(peer_id) = peer_id.parse() {
+							acc.push(peer_id, addrs.into_iter().collect::<HashSet<_>>());
+						}
+						acc
+					},
+				);
+				(protocol, cache)
+			})
+			.collect();
+
+		Self { paths: Arc::new(paths), flushed_at: Instant::now(), protocols, busy: None }
+	}
+
+	pub fn report_peer_addr(
+		&mut self,
+		peer_id: &PeerId,
+		protocol: impl AsRef<[u8]>,
+		addr: &Multiaddr,
+	) {
+		let protocol = String::from_utf8(protocol.as_ref().to_owned()).expect(
+			"According to the `crate::discovery::protocol_name_from_protocol_id` \
+					and `<ProtocolId as AsRef<str>>` it's a correct UTF-8 string",
+		);
+
+		let entries = self
+			.protocols
+			.entry(protocol)
+			.or_insert_with(|| LruCache::new(PEER_ADDRS_CACHE_SIZE));
+		if let Some(peer_addrs) = entries.get_mut(peer_id) {
+			peer_addrs.insert(addr.to_owned());
+		} else {
+			entries.push(peer_id.to_owned(), [addr.to_owned()].into_iter().collect());
+		}
+	}
+
+	pub fn peer_addrs<'a>(
+		&'a mut self,
+		peer_id: &'a PeerId,
+		protocols: impl IntoIterator<Item = impl AsRef<[u8]> + 'a>,
+	) -> impl Iterator<Item = &'a Multiaddr> {
+		let protocols = protocols.into_iter().collect::<Vec<_>>();
+
+		self.protocols
+			.iter_mut()
+			.filter_map(move |(protocol, entries)| {
+				if protocols.iter().any(|p| p.as_ref() == protocol.as_bytes()) {
+					Some(entries)
+				} else {
+					None
+				}
+			})
+			.flat_map(|entries| entries.get(peer_id).into_iter())
+			.flat_map(IntoIterator::into_iter)
+	}
+
+	pub fn poll(&mut self, cx: &mut Context) -> Poll<Never> {
```

Comment:
This is an interesting design; so basically it's a future (but without `Future` being implemented) that never returns, but when polled does some work? Have you considered maybe `tokio::spawn`ing a task, put a `loop` into it, and have it wait for [`tokio::sync::Notify`](https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html)? Or even just do it without a `Notify`, since you have the `FLUSH_INTERVAL` anyway so it's naturally periodical? We could then get rid of the manual juggling of the `busy` future and just `await` on it.

Reply:
@bkchr, I don't think that interacting with the filesystem synchronously is okay within the `poll` of networking behaviour.


@koute,
I thought of spawning a future, but that would bring up:
  * the necessity for the child-task to watch for its parent and terminate once the parent is gone;
  * the necessity for the parent to watch for its child and probably restart it if it terminates prematurely.
 
Solving just the latter would produce a small (and bad) actor system. I decided to remain within the `poll` of the behaviour.

Reply:
> I don't think that interacting with the filesystem synchronously is okay within the poll of networking behaviour.

Honestly, considering how rarely this triggers and how small the files are it should be fine. (...unless we're on Windows, where it's commonplace that arbitrary programs hook into Windows' filesystem layer and install IO hooks; this is why antivirus programs have such a reputation of slowing everything down)

And in this case the implementation also has one extra drawback (although I'd guess it's harmless in practice) - it will only make progress when it is polled by the networking code, *and* if it can't make any progress it will be unnecessarily polled (since the `busy` future's polling status won't get propagated to the outer stack frame hence the task won't get parked). So all in all it does feel a little hacky to me since it's not how futures are usually supposed to be used. (:

Anyway, what I would do is probably something like this, assuming we want to periodically flush the data to disk: (not an exact implementation, pseudocode-ish)

```rust
let state = Arc::new(...);
let state_weak = Arc::downgrade(&state);
tokio::task::spawn(async move {
    while let Some(state) = state_weak.upgrade() {
        let blob = serde_json::to_vec(&state.lock()).expect("serializing JSON doesn't fail");
        if let Err(error) = write_to_file_atomic(&path, &blob).await {
            // ...
        }
        tokio::time::sleep(FLUSH_INTERVAL).await;
    }
});
```

So what you have here is something which:
- Automatically folds down if the parent object is destroyed (`Weak::upgrade` will return `None`)
- There's no need for the parent object to monitor this if it terminates prematurely since the code's so simple that its obvious that it *can't* terminate. (...unless JSON serialization fails, but [that can only happen in very specific circumstances AFAIK](https://www.greyblake.com/blog/when-serde-json-to-string-fails/)). Not that it matters that much anyway, because 1) this is not really critical functionality, and 2) we already run a lot more code which is a lot more complex in tasks which will abort the whole process if they terminate, so this much is no big deal
- It doesn't block the network path with IO. (But with this little IO it probably doesn't make a huge difference.)
- All of the manual juggling with the futures and polling and state tracking (`flushed_at` and `busy`) is gone, and the object doesn't have to be explicitly polled anymore (or in other words, the flushing's automatic, so there's no chance that you will forget to poll it nor that it will accidentally get lost in a future refactoring).

A simpler API, and a simpler implementation.

But it's up to you. (: